pkg/controller/route/route_controller.go

```diff
@@ -137,34 +138,49 @@ func (rc *RouteController) reconcileNodeRoutes() error {
 }
 
 func (rc *RouteController) reconcile(nodes []*v1.Node, routes []*cloudprovider.Route) error {
-	// nodeCIDRs maps nodeName->nodeCIDR
-	nodeCIDRs := make(map[types.NodeName]string)
+	// for each node a map of podCIDRs and their created status
+	var l sync.Mutex
+	nodeRoutesStatuses := make(map[types.NodeName]map[string]bool)
 	// routeMap maps routeTargetNode->route
-	routeMap := make(map[types.NodeName]*cloudprovider.Route)
+	routeMap := make(map[types.NodeName][]*cloudprovider.Route)
 	for _, route := range routes {
 		if route.TargetNode != "" {
-			routeMap[route.TargetNode] = route
+			routeMap[route.TargetNode] = append(routeMap[route.TargetNode], route)
 		}
 	}
 
 	wg := sync.WaitGroup{}
 	rateLimiter := make(chan struct{}, maxConcurrentRouteCreations)
+	// searches existing routes by node for a matching route
 
 	for _, node := range nodes {
 		// Skip if the node hasn't been assigned a CIDR yet.
-		if node.Spec.PodCIDR == "" {
+		if len(node.Spec.PodCIDRs) == 0 {
 			continue
 		}
 		nodeName := types.NodeName(node.Name)
-		// Check if we have a route for this node w/ the correct CIDR.
-		r := routeMap[nodeName]
-		if r == nil || r.DestinationCIDR != node.Spec.PodCIDR {
-			// If not, create the route.
+		l.Lock()
+		nodeRoutesStatuses[nodeName] = make(map[string]bool)
+		l.Unlock()
+		// for every node, for every cidr
+		for _, podCIDR := range node.Spec.PodCIDRs {
+			// we add it to our nodeCIDRs map here because add and delete go routines run at the same time
+			l.Lock()
+			nodeRoutesStatuses[nodeName][podCIDR] = false
+			l.Unlock()
+			// ignore if already created
+			if hasRoute(routeMap, nodeName, podCIDR) {
+				l.Lock()
+				nodeRoutesStatuses[nodeName][podCIDR] = true // a route for this podCIDR is already created
+				l.Unlock()
+				continue
+			}
+			// if we are here, then a route needs to be created for this node
 			route := &cloudprovider.Route{
 				TargetNode:      nodeName,
-				DestinationCIDR: node.Spec.PodCIDR,
+				DestinationCIDR: podCIDR,
 			}
-			nameHint := string(node.UID)
+			nameHint := string(node.UID) + cidrtoRfc1035(podCIDR)
```

Comment:
So after looking at a bunch of provider implementations, it seems like changing `nameHint` shouldn't matter since we don't rely on it for listing or deleting routes afterwards. Will check some of the out-of-tree providers to be sure.